### PR TITLE
[X86] Don't convert local function foo in the same section to foo(%rip) when the offset is near INT32_MIN

### PR DESCRIPTION
**Pull request by MaskRay**
```
define internal void @foo() {
  ret void
}
define i64 @main() {
  ret i64 add (i64 ptrtoint (ptr @foo to i64), i64 -2147483626)
}
```

When `foo` is a local symbol, `foo` and `main` are in the same section,
and `offset` is near INT32_MIN, referencing `foo+offset` in `main` with
RIP-relative addressing needs `leaq .text+offset1(%rip), %rax` where
`offset1 < offset`, and `offset1` might underflow.
(https://discourse.llvm.org/t/arithmetic-referencing-dso-local-function-causes-compilation-error-on-linux-x64/80033):

Don't use RIP-relative addressing if the negative offset is near
INT32_MIN. Arbitrarily reuse the magic number in isOffsetSuitableForCodeModel to
guard against the edge case when `address(current_instruction)-foo < 4GiB-16MiB`.
If the difference is larger than 4GiB-16MiB, `ret i64 add (i64 ptrtoint
(ptr @foo to i64), i64 -2**32+256MiB)` would still cause the assembly
issue, such cases are unrealistic.

### CODE REVIEW

File: llvm/lib/Target/X86/X86ISelDAGToDAG.cpp [hunk 1]
```diff
@@ -1795,6 +1795,13 @@ bool X86DAGToDAGISel::foldOffsetIntoAddress(uint64_t Offset,
         !X86::isOffsetSuitableForCodeModel(Val, M,
                                            AM.hasSymbolicDisplacement()))
       return true;
+    // When Sym is a local symbol in ELF, leaq Sym+Offset(%rip) in non-large
+    // code models generates a relocation referencing the section symbol. Don't
+    // fold offsets that are too closer to INT32_MIN, as the addend could
+    // underflow.
+    if (Val < INT32_MIN + 256 * 1024 * 1024 && M != CodeModel::Large &&
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:1802]: What led you to the 256 * 1024 * 1024 magic number?

**Comment by jyknight** [comment]
I don't think we should try to fix this. There's no way to know at compile time how far the referenced symbol will be from RIP, and thus how big an offset would be too big.

**Comment by MaskRay** [comment]
> I don't think we should try to fix this. There's no way to know at compile time how far the referenced symbol will be from RIP, and thus how big an offset would be too big.

The previous analysis was incorrect. Sorry, wrote the patch very late last night.

The latest update keeps the special case for very limited cases.

> @RKSimon What led you to the 256 * 1024 * 1024 magic number? 

The description has been updated to state that this is arbitrary (like the 16*1024*1024 in `isOffsetSuitableForCodeModel`).
When `foo` and `main` are in the same section, we support `main-foo` up to 256MiB.

If main-foo is larger than 256MiB, and we have `ret i64 add (i64 ptrtoint (ptr @foo to i64), i64 -2**32+256MiB)`, we would still run into an issue, but hopefully such cases are very unrealistic.

LTO builds default to -ffunction-sections, and a section in a relocatable file can unlikely be larger than 256MiB.

**Comment by jyknight** [comment]
But that's only _one_ of the kinds of errors you can get from such code.

We'll also emit a PC32 relocation -- with the user-specified addend included -- for symbols that are _NOT_ in the same section. It won't be an assembler error, then, but will turn into a linker error. A binary can easily be larger than 256MB, in which case this breaks again.

I really don't think magic numbers are the right answer here. We either need to formalize this sort of code as invalid, or else modify codegen to not use a 32-bit-limited fixup/relocation for code like this.

**Comment by MaskRay** [comment]
> But that's only _one_ of the kinds of errors you can get from such code.

The other kind of errors are prevented by other code, e.g. `isOffsetSuitableForCodeModel` has `return Offset < 16 * 1024 * 1024`.

> We'll also emit a PC32 relocation -- with the user-specified addend included -- for symbols that are _NOT_ in the same section. It won't be an assembler error, then, but will turn into a linker error. A binary can easily be larger than 256MB, in which case this breaks again.

```
define internal void @foo() {
  ret void
}
define i64 @main() {
  ret i64 add (i64 ptrtoint (ptr @foo to i64), i64 -1879048192)
}
define i64 @bar() {
  ret i64 add (i64 ptrtoint (ptr @foo to i64), i64 1879048192)
}
```

When `foo` and `main` are in different sections, MC will emit a relocation.
There won't be a linker error, even if `--emit-relocs` is used.
(When ld --emit-relocs converts `.text.foo - 0x70000004 to .text + addend`, the `addend` will be larger than -0x70000004.)

```
% readelf -Wr r.o
Relocation section '.rela.text.main' at offset 0xd0 contains 1 entry:
    Offset             Info             Type               Symbol's Value  Symbol's Name + Addend
0000000000000003  0000000200000002 R_X86_64_PC32          0000000000000000 .text.foo - 70000004
```

---

Even when `foo` and `bar` are in the same section, we will not fold the large displacement (0x70000000) into leaq
due to `return Offset < 16 * 1024 * 1024` in `isOffsetSuitableForCodeModel`.

```asm
        movl    $1879048192, %eax               # imm = 0x70000000
        leaq    foo(%rax), %rax
        retq
```


> I really don't think magic numbers are the right answer here. We either need to formalize this sort of code as invalid, or else modify codegen to not use a 32-bit-limited fixup/relocation for code like this.

The magic number is about the input section section size, instead of the output section size.
If the leaq optimization of a function symbol with a negative offset like -0x70000000 isn't that useful, I could also drop it completely.

**Comment by MaskRay** [comment]
> The other kinds of errors are prevented by other code, e.g. `isOffsetSuitableForCodeModel` has `return Offset < 16 * 1024 * 1024`.

@jyknight Does the magic offset used in a similar place make me accept this approach more?

I am happy to change that INT_MIN+256MiB to 16MiB to have consistent magic numbers.

**Comment by MaskRay** [comment]
Friendly ping :)

**Comment by MaskRay** [comment]
> > The other kinds of errors are prevented by other code, e.g. `isOffsetSuitableForCodeModel` has `return Offset < 16 * 1024 * 1024`.
> 
> @jyknight Does the magic offset used in a similar place make you accept this approach more? :)
> 
> I am happy to change that INT_MIN+256MiB to 16MiB to have consistent magic numbers.

Friendly ping :)

**Comment by RKSimon** [comment]
Let's at least use the same offset as used in isOffsetSuitableForCodeModel - ideally add comments to both locations saying that they are matching estimates (you can pull out the magic number into one location if you want but its not necessary).

**Comment by MaskRay** [comment]
Ping

**Review by RKSimon — APPROVED** [review]
LGTM as a sticking plaster, but it'd be much better if we can come up with better solutions for both compile and link time cases.

**Pull request merged**